pkg/patterns/declarative/pkg/watch/dynamic.go

```diff
@@ -0,0 +1,81 @@
+package watch
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+func NewDynamicWatch(config rest.Config) (*dynamicWatch, chan event.GenericEvent, error) {
+	dw := &dynamicWatch{events: make(chan event.GenericEvent)}
+
+	restMapper, err := apiutil.NewDiscoveryRESTMapper(&config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	client, err := dynamic.NewForConfig(&config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	dw.restMapper = restMapper
+	dw.config = config
+	dw.client = client
+	return dw, dw.events, nil
+}
+
+type dynamicWatch struct {
+	config     rest.Config
+	client     dynamic.Interface
+	restMapper meta.RESTMapper
+	events     chan event.GenericEvent
+}
+
+func (dw *dynamicWatch) newDynamicClient(gvk schema.GroupVersionKind) (dynamic.ResourceInterface, error) {
+	mapping, err := dw.restMapper.RESTMapping(gvk.GroupKind(), gvk.Version)
```

Comment:
What is this magic?  I wish I had known about this when I was hand coding this mapping for the Nth time ;-)